[move-book] Perform maintenance, draft named addresses (#9195)
diem/diem: language/documentation/book/src/address.md

You can use runtime address values (values of type `address`) to access resources at that address. You *cannot* access modules at runtime via address values.

## Addresses and Their Syntax

Addresses come in two flavors, named or numerical. The syntax for a named address follows the
same rules for any named identifier in Move. The syntax of a numerical address is not restricted
to hex-encoded values, and any valid [`u128` numerical value](./integers.md) can be used as an
address value, e.g., `42`, `0xCAFE`, and `2021` are all valid numerical address
literals.

To distinguish when an address is being used in an expression context or not, the
syntax when using an address will differ slightly depending on the context it's
being used in:

> * When an address is used as an expression the address must be prefixed by the `@` character, i.e., `@`[`<numeric_value>`](./integers.md)` or `@<named_address_identifier>`.

Reviewer: 👍

Reviewer: This looks much better! Do we link to the module section here at all?

Author: We don't here, but I just updated below in the part on named addresses to point to the module section.

You can use runtime address values (values of type `address`) to access resources at that address. You *cannot* access modules at runtime via address values.

## Addresses and Their Syntax

Addresses come in two flavors, named or numerical. The syntax for a named address follows the
same rules for any named identifier in Move. The syntax of a numerical address is not restricted
to hex-encoded values, and any valid [`u128` numerical value](./integers.md) can be used as an
address value, e.g., `42`, `0xCAFE`, and `2021` are all valid numerical address
literals.

To distinguish when an address is being used in an expression context or not, the

> syntax when using an address will differ slightly depending on the context it's

Reviewer 2: ```suggestion
syntax when using an address differs depending on the context where it is used.
```

(also replacing the next line)

being used in:

> * When an address is used as an expression the address must be prefixed by the `@` character, i.e., `@`[`<numeric_value>`](./integers.md)` or `@<named_address_identifier>`.

Reviewer 2: There's an interesting switch in terminology here. I don't know if it is intentional. Earlier you used the term "numerical address", and I was about to suggest changing to "numeric address" but then looked up both in the dictionary and decided that there was no clear reason to change. But, now you're using `numeric_value`.... For consistency, shouldn't that be `numerical_value`? Or is there some subtle distinction that you're making between these?

Author: Non-intentional and great spot. I think for consistency this should definitely be `numerical_value` (it's even referenced like that further above in this doc!).

* Outside of expression contexts, the address may be written without the leading `@` character, i.e., [`<numeric_value>`](./integers.md) or `<named_address_identifier>`.

In general, you can think of `@` as an operator that takes an address from being a namespace item to being an expression item.

## Named Addresses

> As mentioned above, Move supports the concept of named addresses. Named

Reviewer 2: Kill this first sentence. It was already mentioned above, and this sentence adds no value.

addresses are a feature that allow identifiers to be used in place of numerical
values in any spot where addresses are used, and not just at the value level.
Named addresses are declared and bound as top level elements (outside of modules and
scripts) in Move Packages, or passed as arguments to the Move compiler.

> Named addresses only exist at the source language level and during compilation

Reviewer 2: ```suggestion
Named addresses only exist at the source language level
```

and will be fully substituted for their value at the bytecode level. Because of

> this, their value should be treated opaquely if at all possible -- and the

Reviewer 2: I have multiple issues with this "Because...." sentence (e.g., "value" should be plural, when it is not possible?) I think it would be best to just skip this and cut to the chase in the following sentence:

Because of this, modules and module members _must_ be accessed....

You can use runtime address values (values of type `address`) to access resources at that address. You *cannot* access modules at runtime via address values.

## Addresses and Their Syntax

Addresses come in two flavors, named or numerical. The syntax for a named address follows the
same rules for any named identifier in Move. The syntax of a numerical address is not restricted

> to hex-encoded values, and any valid numerical literal can be used as an address value.

Reviewer: I might give a little grammar breakdown or something here to show the different cases

To distinguish when an address is being used in an expression context or not, the
syntax when using an address will differ slightly depending on the context it's
being used in:
* When an address is used as an expression the address must be prefixed by the `@` character, i.e., `@value`.

> * Outside of expression contexts, the address may be written without the leading `@` character, i.e., `value`.

Reviewer: might want to link somewhere else
